### ISSUE TITLE
Make Capsule Frame before Headers an H3_FRAME_UNEXPECTED connection error

### ISSUE THREAD

**Issue by lnicco** [ietf-wg-masque/draft-ietf-masque-h3-datagram#73]
The HTTP/3 RFC states 
> Receipt of an invalid sequence of frames MUST be treated 
> as a connection error of type H3_FRAME_UNEXPECTED;

The current draft instead recommends closing the stream rather than the connection and with the generic H3_GENERAL_PROTOCOL_ERROR

According to https://datatracker.ietf.org/doc/html/draft-ietf-quic-http-34#section-8.1 H3_GENERAL_PROTOCOL_ERROR is to be used when a more specific error code isn't available, but here we have clearly a more specific error available.

Also, what is the rationale for closing just the stream and not the whole connection?

Furthermore, a more general question:
right now the above recommendation only exists for REGISTER_DATAGRAM_NO_CONTEXT and REGISTER_DATAGRAM_CONTEXT capsule types. Would it be fair to say that no Capsule frame is allowed before headers instead?

**Comment by LPardue** [comment]
Thanks.

I think it would be for us to clarify

a) strict ordering rules of capsule amongst HTTP message frames.
b) strict ordering rules of capsule types.

For (a), I'm inclined to state a top-level requirement that CAPSULE frames MUST NOT arrive before the first HEADERS on a request stream. Violation of that is connection error of type H3_FRAME_UNEXPECTED. 

For (b), I'm slightly less sure. We clearly have some ordering requirements. The H3 base spec treats misordering as a critical failure. We treat it here as a stream error. I think we could make things more similar to H3. For instance, I can't think of a good reason right now why we are so loose with handling of datagram capsules; if they arrive on a stream before a register capsule, that smells broken.

**Comment by DavidSchinazi** [comment]
Banning all capsules before headers sounds reasonable to me.

Strict ordering of capsules is required on the sender, but enforcing that on the receiver isn't always easy so I don't think we can mandate it.

I honestly don't care if it's a stream or connection error, either works for me.

**Comment by LPardue** [comment]
Happy to write the PR for this. Right now I'm thinking we cam make (b) a more specific error code too. I think H3_FRAME_UNEXPECTED is good enough for capsule types but we could always define a H3_CAPSULE _UNEXPECTED is people felt strongly.

**Comment by DavidSchinazi** [comment]
Do you have an example of where (b) is useful? In other words, what's a sequence of events that leads a receiver to detect a capsule illegal ordering?

**Comment by afrind** [comment]
That last sequence probably can't be illegal, since it is possible when forwarding through an intermediary with and without datagram support.  Eg: Client <-> Intermediary supports native datagram and Intermediary <-> Server only supports datagram CAPSULE.  Since the intermediary only needs to forward and not inspect, the datagram CAPSULE can arrive before the register_datagram_no_context CAPSULE.

**Comment by LPardue** [comment]
That's what I thought initially, but the spec presently says

> due to reordering, an endpoint that receives a DATAGRAM frame before receiving either a REGISTER_DATAGRAM_CONTEXT capsule or a REGISTER_DATAGRAM_NO_CONTEXT capsule MUST NOT treat it as an error, it SHALL instead drop the DATAGRAM frame silently, or buffer it temporarily while awaiting a REGISTER_DATAGRAM_NO_CONTEXT capsule or the corresponding REGISTER_DATAGRAM_CONTEXT capsule

By my read, an intermediary that receives a DATAGRAM frame and wants to convert it to a capsule needs to wait for the register capsule.

**Comment by afrind** [comment]
The spec doesn't define endpoint.  Is an intermediary an endpoint?

**Comment by LPardue** [comment]
Need to nail that down. Having different sending rules for clients and "intermediary doing client behaviour" is a recipe for illness, the server can't know what the sender thinks it is.

**Comment by DavidSchinazi** [comment]
An endpoint is the opposite of an intermediary. If that's not obvious then we should spell it out in the draft. Intermediaries do not parse register capsules.

**Comment by LPardue** [comment]
I think the parsing stuff is clear. What's a bit unclear is the message forwarding requirements. The HTTP definition [1] of intermediary allows some variance. In the gateway case, it's stated that

>  an HTTP-to-HTTP gateway that wishes to interoperate with
   third-party HTTP servers needs to conform to user agent requirements
   on the gateway's inbound connection.

Allowing deviation is permitted but might cause interop problems.

[1] - https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-semantics-16#section-3.7

**Comment by DavidSchinazi** [comment]
I'm not sure what you're suggesting in terms of changes to the draft @LPardue, could you elaborate?

**Comment by LPardue** [comment]
The specific suggestion in this case is, disallow intermediaries from sending datagram capsules before register capsules. This might make Alan sad because an intermediary would need to buffer DATAGRAM frames and only forward them as capsules as a register is seen. Note that it would not require the intermediary to act on the register capsule, just enforce sequencing rules between datagrams and streams

**Comment by DavidSchinazi** [comment]
I'd rather we didn't add that suggestion. It increases complexity for intermediaries and I'm not sure what problem it solves?

**Comment by LPardue** [comment]
That's fine. I'd be happy with not adding that. However, to make it clear what the spec is allowing, we should make a note in Section 4.4 that the ordering/buffering requirements of DATGRAM frames and register capsules (from sec 3) do not apply to intermediaries when forwarding. That's tangential to Luca's OP so we can spin it off separately.

**Comment by DavidSchinazi** [comment]
I agree, that sounds like a good resolution. Can you file an issue please?
